jax/interpreters/numpy_eval.py

```diff
@@ -0,0 +1,597 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from contextlib import contextmanager
+from typing import Any, Dict, Callable
+import builtins
+import itertools
+import numpy as np
+import opt_einsum
+import scipy.special
+
+from .. import config, core, dtypes, xla, curry, lax
+from ..interpreters import xla
+
+_slice = builtins.slice
+_min = builtins.min
+_map = builtins.map
+
+np_impl: Dict[core.Primitive, Callable[..., Any]] = {}
+
+class NumpyEvalTrace(core.Trace):
+  def pure(self, x): return x
+  lift = sublift = pure
+
+  def process_primitive(self, primitive, tracers, params):
+    impl = np_impl.get(primitive)
+    if impl is None:
+      raise NotImplementedError(f"No original numpy implementation for "
+                                f"'{primitive}' primitive found.")
+    return impl(*tracers, **params)
+
+  def process_call(self, primitive, f, tracers, params):
+    return f.call_wrapped(*tracers)
+  process_map = process_call
+
+@contextmanager
+def numpy_eval():
+  """
```

Comment:
Can you document whether the use of functions decorated with `numpy_eval` is thread safe? I'm guessing the answer is yes?

Reply:
@shoyer Yes, and yes, done!